### PR TITLE
curve spec swap and add_liquidity

### PR DESCRIPTION
**Pull request by danoctavian**
Adds test for curve swaps

Fixes MockST_ETH.sol to be rebasing like STETH.

Fixes Provider rate for STETH (1:1)

### CODE REVIEW

File: test/unit/mocks/MockST_ETH.sol [hunk 1]
```diff
@@ -3,40 +3,70 @@ pragma solidity ^0.8.24;
 
 // https://github.com/lidofinance/lido-dao/blob/master/contracts/0.4.24/StETH.sol
 import {ERC20} from "src/Common.sol";
+import {IStETH} from "src/interface/external/lido/IStETH.sol";
 
-contract MockSTETH is ERC20 {
-    uint256 private _pooledEthPerShare = 1e18; // Start with 1:1 ratio
-    mapping(address => uint256) public shares;
+contract MockSTETH is IStETH, ERC20 {
```
Comment (danoctavian) [inline test/unit/mocks/MockST_ETH.sol:8]: fixed this mock so it behaves like steth.

before this, the stETH value was not 1:1 with ETH

This has rebasing working, and can accrue rewards through the function addRewards(). The function does not change the stETH:ETH ratio but increases everyone's stETH balance.

File: src/module/Provider.sol [hunk 2]
```diff
@@ -23,7 +23,7 @@ contract Provider is IProvider {
         } else if (asset == MC.BUFFER) {
             return IERC4626(MC.BUFFER).previewRedeem(1e18);
         } else if (asset == MC.STETH) {
-            return IStETH(MC.STETH).getPooledEthByShares(1e18);
+            return 1e18;
```
Comment (danoctavian) [inline src/module/Provider.sol:26]: for STETH the ratio here is 1:1

for WSTETH we need another entry

File: test/mainnet/curve.spec.sol [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+            vm.startPrank(PROCESSOR);
+            vault.processor(targets, values, callData);
+            vm.stopPrank();
+        }
+
+        // Assert ETH balance is reduced and LP tokens were received
+        assertApproxEqAbs(address(vault).balance, 0, 2);
+        assertApproxEqAbs(IERC20(MC.STETH).balanceOf(address(vault)), 0, 2);
+        uint256 lpBalance = IERC20(pool.lp_token()).balanceOf(address(vault));
+        assertGt(lpBalance, minLpOut);
```
Comment (danoctavian) [inline test/mainnet/curve.spec.sol]: important note here: if we support add_liquidity then we need the LP pool as a strategy and a rate for it

File: src/module/Provider.sol [hunk 4]
```diff
@@ -23,6 +23,8 @@ contract Provider is IProvider {
         } else if (asset == MC.BUFFER) {
             return IERC4626(MC.BUFFER).previewRedeem(1e18);
         } else if (asset == MC.STETH) {
+            return 1e18;
+        } else if (asset == MC.WSTETH) {
             return IStETH(MC.STETH).getPooledEthByShares(1e18);
```
Comment (danoctavian) [inline src/module/Provider.sol:28]: getPooledEthByShares provides the rate for WSTETH

File: test/mainnet/helpers/SetupVault.sol [hunk 5]
```diff
@@ -89,7 +91,9 @@ contract SetupVault is Test, MainnetActors, Etches {
         vault.addStrategy(MC.YNETH, 18);
         vault.addStrategy(MC.YNLSDE, 18);
 
-        vault.setBuffer(MC.BUFFER);                                                                  
+        vault.setBuffer(MC.BUFFER);     
+
+        configureCurveActions(vault);
```
Comment (xhad) [inline test/mainnet/helpers/SetupVault.sol]: Kindly move this test specific configuration to the curve test so the global Setup is generic and simple

File: src/module/Provider.sol (hunk 2 above)
Comment (xhad) [reply src/module/Provider.sol:26]: ok good

File: test/mainnet/curve.spec.sol (hunk 3 above)
Comment (xhad) [reply test/mainnet/curve.spec.sol]: yeah, let's sync on this arch topic today

File: test/mainnet/helpers/SetupVault.sol [hunk 6]
```diff
@@ -98,6 +102,81 @@ contract SetupVault is Test, MainnetActors, Etches {
...
+        // Get ethSteth pool from registry
+        ICurveRegistry registry = ICurveRegistry(MC.CURVE_REGISTRY);
+        address ethStethPool = registry.find_pool_for_coins(MC.ETH, MC.STETH);
+
+        // Get ynETHWstETH pool from two crypto factory
+        address ynETHWstETHPool = ICurveRegistry(MC.CURVE_TWOCRYPTO_FACTORY).find_pool_for_coins(MC.YNETH, MC.WSTETH);
+
+        // Add curve pools to array
+        address[] memory curvePools = new address[](1);
+        curvePools[0] = ethStethPool;
```
Comment (xhad) [inline test/mainnet/helpers/SetupVault.sol]: can remove the two lines above as the array is created below

File: test/mainnet/helpers/SetupVault.sol [hunk 7]
```diff
@@ -7,6 +7,8 @@ import {TimelockController as TLC} from "src/Common.sol";
 import {MainnetActors} from "script/Actors.sol";
 import {MainnetContracts as MC} from "script/Contracts.sol";
 import {Etches} from "test/mainnet/helpers/Etches.sol";
+import {ICurveRegistry} from "src/interface/external/curve/ICurveRegistry.sol";
+import {ICurvePool} from "src/interface/external/curve/ICurvePool.sol";
```
Comment (xhad) [inline test/mainnet/helpers/SetupVault.sol]: ICurvePool appears to be unused

**Review by xhad — APPROVED** [review]

**Pull request merged**